File: src/main/java/me/pcasaes/hexoids/core/domain/model/Players.java
package me.pcasaes.hexoids.core.domain.model;

import me.pcasaes.hexoids.core.domain.config.Config;
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndex;
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndexFactory;
import me.pcasaes.hexoids.core.domain.utils.MathUtil;
import me.pcasaes.hexoids.core.domain.vector.Vector2;
import pcasaes.hexoids.proto.BoltExhaustedEventDto;
import pcasaes.hexoids.proto.BoltFiredEventDto;
import pcasaes.hexoids.proto.BoltsAvailableCommandDto;
import pcasaes.hexoids.proto.CurrentViewCommandDto;
import pcasaes.hexoids.proto.DirectedCommand;
import pcasaes.hexoids.proto.Dto;
import pcasaes.hexoids.proto.PlayerDestroyedEventDto;
import pcasaes.hexoids.proto.PlayerJoinedEventDto;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

/**
 * The collection of Players.
 */
public class Players implements Iterable<Player> {

    static Players create(Bolts bolts, Clock clock, ScoreBoard scoreBoard, Barriers barriers, PlayerSpatialIndexFactory spatialIndexFactory) {
        return new Players(bolts, clock, scoreBoard, barriers, spatialIndexFactory);
    }

    /**
     * All players are maintained in this map
     */
    private final Map<EntityId, Player> playerMap = new HashMap<>(5000, 0.5f);

    /**
     * Only players connected to this node are maintained in this set
     */
    private final Set<EntityId> playerServerUpdateSet = new HashSet<>(5000, 0.5f);

    private final Bolts bolts;

    private final Clock clock;

    private final ScoreBoard scoreBoard;

    private final Barriers barriers;

    private final PlayerSpatialIndexFactory spatialIndexFactory;

    private Players(Bolts bolts, Clock clock, ScoreBoard scoreBoard, Barriers barriers, PlayerSpatialIndexFactory spatialIndexFactory) {
        this.bolts = bolts;
        this.clock = clock;
        this.scoreBoard = scoreBoard;
        this.barriers = barriers;
        this.spatialIndexFactory = spatialIndexFactory;
    }

    /**
     * If the player hasn't been created will do so and return the player
     *
     * @param id player's id.
     * @return If created returns the player
     */
    public Optional<Player> createPlayer(EntityId id) {
        if (playerMap.containsKey(id)) {
            return Optional.empty();
        }
        return Optional.of(createOrGet(id));
    }

    /**
     * Returns a player, creating them if they don't exist.
     *
     * @param id the player's id.
     * @return the player
     */
    public Player createOrGet(EntityId id) {
        return playerMap.computeIfAbsent(id, this::create);
    }

    /**
     * Returns a specific player if they exist.
     * <p>
     *
     * @param id player's identifier
     * @return Returns the player if they exist
     */
    public Optional<Player> get(EntityId id) {
        return Optional.ofNullable(playerMap.get(id));
    }

    private Optional<Player> get(UUID id) {
        return get(EntityId.of(id));
    }

    private Player create(EntityId id) {
        requestCurrentView(id);
        return Player.create(id, this, this.bolts, this.barriers, this.clock, this.scoreBoard);
    }

    /**
     * Requests the game's current view to be sent to a specific player
     *
     * @param requesterId player to have the current view sent to.
     */
    public void requestCurrentView(EntityId requesterId) {
        CurrentViewCommandDto.Builder currentViewBuilder = CurrentViewCommandDto.newBuilder()
                .addAllBarriers(StreamSupport.stream(barriers.spliterator(), false)
                        .map(Barrier::toDto)
                        .collect(Collectors.toList())
                )
                .setBoltsAvailable(BoltsAvailableCommandDto.newBuilder().setAvailable(Config.get().getMaxBolts()));

        playerMap
                .values()
                .stream()
                .map(Player::toDtoIfJoined)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .forEach(currentViewBuilder::addPlayers);

        DirectedCommand.Builder builder = DirectedCommand.newBuilder()
                .setPlayerId(requesterId.getGuid())
                .setCurrentView(currentViewBuilder);

        GameEvents.getClientEvents().dispatch(
                Dto.newBuilder()
                        .setDirectedCommand(builder)
                        .build()
        );
    }

    /**
     * Iterator of players
     *
     * @return
     */
    @Override
    public Iterator<Player> iterator() {
        return playerMap
                .values()
                .iterator();
    }

    /**
     * Stream of players
     *
     * @return
     */
    public Stream<Player> stream() {
        return StreamSupport.stream(spliterator(), false);
    }

    void joined(PlayerJoinedEventDto event) {
        Player player = createOrGet(EntityId.of(event.getPlayerId()));
        player.joined(event);
    }

    private void left(EntityId playerId) {
        Player player = createOrGet(playerId);
        playerMap.remove(playerId);
        playerServerUpdateSet.remove(playerId);
        player.left();
    }

    void consumeFromJoinTopic(DomainEvent domainEvent) {
        if (domainEvent.getEvent() == null) {
            left(EntityId.of(domainEvent.getKey()));
        } else if (domainEvent.getEvent().hasPlayerJoined()) {
            joined(domainEvent.getEvent().getPlayerJoined());
        }
    }

    void consumeFromPlayerActionTopic(DomainEvent domainEvent) {
        if (domainEvent.getEvent() != null && domainEvent.getEvent().hasPlayerMoved()) {
            get(domainEvent.getKey())
                    .ifPresent(p -> p.moved(domainEvent.getEvent().getPlayerMoved()));
        }
        if (domainEvent.getEvent() != null && domainEvent.getEvent().hasPlayerDestroyed()) {
            get(domainEvent.getKey())
                    .ifPresent(p -> handleDestroyed(p, domainEvent.getEvent().getPlayerDestroyed()));
        }
        if (domainEvent.getEvent() != null && domainEvent.getEvent().hasPlayerSpawned()) {
            get(domainEvent.getKey())
                    .ifPresent(p -> p.spawned(domainEvent.getEvent().getPlayerSpawned()));
        }
    }

    void handleDestroyed(Player player, PlayerDestroyedEventDto playerDestroyedEvent) {
        handleShockwave(player);
        player.destroyed(playerDestroyedEvent);
    }

    /**
     * When a player is destroyed we generate a shockwave that pushes nearby players away
     * from the destroyed players last position.
     *
     * @param fromPlayer
     */
    void handleShockwave(Player fromPlayer) {
        final var dist = Config.get().getPlayerDestroyedShockwaveDistance();
        if (dist <= 0F) {
            return;
        }
        Vector2 fromPlayerPosition = Vector2.fromXY(fromPlayer.getX(), fromPlayer.getY());
        getSpatialIndex()
                .search(fromPlayer.getX(), fromPlayer.getY(), fromPlayer.getX(), fromPlayer.getY(), dist)
                .forEach(nearPlayer -> {
                    if (nearPlayer != fromPlayer) {
                        Vector2 distanceBetweenPlayers = Vector2
                                .fromXY(nearPlayer.getX(), nearPlayer.getY())
                                .minus(fromPlayerPosition);

                        // rescaled to be between 0.0 and 1.0 inclusive
                        float absMagRescaled = Math.abs(distanceBetweenPlayers.getMagnitude()) / dist;

                        boolean isNotCenteredNorOutOfRange = absMagRescaled <= 1F && absMagRescaled > 0F;
                        if (isNotCenteredNorOutOfRange) {
                            int sign = distanceBetweenPlayers.getMagnitude() < 0F ? -1 : 1;

                            float invertedAbsMag = (1F - absMagRescaled);
                            if (invertedAbsMag > 0.5) {
                                // if near destroyed player let's limit non-linearly (quadratic) the shockwave magnitude
                                // f(x) =   0.75 - (x - 1)^2    : x > 0.5
                                //          x                   : x <= 0.5
                                // at 0.5 the quadratic transform intersects tangentially to f(x) = x
                                invertedAbsMag = 0.75F - MathUtil.square(invertedAbsMag - 1F);
                            }

                            Vector2 move = Vector2
                                    .fromAngleMagnitude(
                                            distanceBetweenPlayers.getAngle(),
                                            sign * invertedAbsMag * dist
                                    );

                            nearPlayer.move(move.getX(), move.getY(), null);
                        }
                    }
                });
    }

    /**
     * This call marks the player is connected on this running instance.
     * This is used to to update server calculated vector positions.
     *
     * @param playerId
     */
    public void connected(EntityId playerId) {
        playerServerUpdateSet.add(playerId);
    }

    void consumeFromPlayerFiredTopic(DomainEvent domainEvent) {
        if (domainEvent.getEvent() != null && domainEvent.getEvent().hasPlayerFired()) {
            BoltFiredEventDto playerFiredEventDto = domainEvent.getEvent().getPlayerFired();
            get(EntityId.of(playerFiredEventDto.getOwnerPlayerId()))
                    .ifPresent(p -> p.fired(domainEvent.getEvent().getPlayerFired()));
        }
    }

    void consumeFromBoltActionTopic(DomainEvent domainEvent) {
        if (domainEvent.getEvent() != null && domainEvent.getEvent().hasBoltExhausted()) {
            BoltExhaustedEventDto event = domainEvent.getEvent().getBoltExhausted();
            get(EntityId.of(event.getOwnerPlayerId()))
                    .ifPresent(Player::boltExhausted);
        }
    }

    /**
     * Updates all connected player models' vector positions.
     *
     * @param timestamp the timestamp to update the players to.
     */
    void fixedUpdate(long timestamp) {
        playerServerUpdateSet
                .stream()
                .map(playerMap::get)
                .filter(Objects::nonNull)
                .forEach(p -> p.fixedUpdate(timestamp));
    }

    /**
     * Return the total number of players in the game.
     * Is weakly consistent and thread safe.
     *
     * @return
     */
    public int getTotalNumberOfPlayers() {
        return this.playerMap.size();
    }

    /**
     * Return the number of players connected this node.
     * Is weakly consistent and thread safe.
     *
     * @return
     */
    public int getNumberOfConnectedPlayers() {
        return this.playerServerUpdateSet.size();
    }

    PlayerSpatialIndex getSpatialIndex() {
        return spatialIndexFactory.get();
    }

    boolean isConnected(EntityId id) {
        return playerServerUpdateSet.contains(id);
    }
}

File: src/main/kotlin/me/pcasaes/hexoids/core/domain/model/Players.kt
package me.pcasaes.hexoids.core.domain.model

import me.pcasaes.hexoids.core.domain.config.Config
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndex
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndexFactory
import me.pcasaes.hexoids.core.domain.model.EntityId.Companion.of
import me.pcasaes.hexoids.core.domain.model.GameEvents.Companion.getClientEvents
import me.pcasaes.hexoids.core.domain.model.physics.Shockwave.Companion.shipExploded
import pcasaes.hexoids.proto.BoltsAvailableCommandDto
import pcasaes.hexoids.proto.CurrentViewCommandDto
import pcasaes.hexoids.proto.DirectedCommand
import pcasaes.hexoids.proto.Dto
import pcasaes.hexoids.proto.PlayerDestroyedEventDto
import pcasaes.hexoids.proto.PlayerJoinedEventDto
import java.util.UUID
import java.util.function.Consumer

/**
 * The collection of Players.
 */
class Players private constructor(
    private val bolts: Bolts,
    private val clock: Clock,
    private val scoreBoard: ScoreBoard,
    private val barriers: Barriers,
    private val physicsQueue: PhysicsQueueEnqueue,
    private val spatialIndexFactory: PlayerSpatialIndexFactory
) : Iterable<Player> {
    /**
     * All players are maintained in this map
     */
    private val playerMap = HashMap<EntityId, Player>(5000, 0.5F)

    /**
     * Only players connected to this node are maintained in this set
     */
    private val playerServerUpdateSet = HashSet<EntityId>(5000, 0.5F)

    private val registeredCurrentViewModifiers = HashMap<EntityId, Consumer<CurrentViewCommandDto.Builder>>()

    /**
     * If the player hasn't been created will do so and return the player
     *
     * @param id player's id.
     * @return If created returns the player
     */
    fun createPlayer(id: EntityId): Player? {
        return if (playerMap.containsKey(id)) {
            null
        } else {
            createOrGet(id)
        }
    }

    /**
     * Returns a player, creating them if they don't exist.
     *
     * @param id the player's id.
     * @return the player
     */
    fun createOrGet(id: EntityId): Player {
        return playerMap.computeIfAbsent(id) { id -> this.create(id) }
    }

    /**
     * Returns a specific player if they exist.
     *
     *
     *
     * @param id player's identifier
     * @return Returns the player if they exist
     */
    fun get(id: EntityId): Player? {
        return playerMap[id]
    }

    private fun get(id: UUID): Player? {
        return get(of(id))
    }

    private fun create(id: EntityId): Player {
        requestCurrentView(id)
        return Player.create(id, this, this.bolts, this.barriers, this.clock, this.scoreBoard)
    }

    fun registerCurrentViewModifier(modifierId: EntityId, builderConsumer: Consumer<CurrentViewCommandDto.Builder>) {
        registeredCurrentViewModifiers.put(modifierId, builderConsumer)
    }

    fun unregisterCurrentViewModifier(modifierId: EntityId) {
        registeredCurrentViewModifiers.remove(modifierId)
    }

    /**
     * Requests the game's current view to be sent to a specific player
     *
     * @param requesterId player to have the current view sent to.
     */
    fun requestCurrentView(requesterId: EntityId) {
        val currentViewBuilder = CurrentViewCommandDto.newBuilder()
            .addAllBarriers(
                barriers.map { it.toDto() }
            )
            .setBoltsAvailable(BoltsAvailableCommandDto.newBuilder().setAvailable(Config.getMaxBolts()))

        registeredCurrentViewModifiers
            .values
            .forEach { c -> c.accept(currentViewBuilder) }

        playerMap
            .values
            .asSequence()
            .map { obj -> obj.toDtoIfJoined() }
            .forEach { value ->
                if (value != null) {
                    currentViewBuilder.addPlayers(value)
                }
            }

        val builder = DirectedCommand.newBuilder()
            .setPlayerId(requesterId.getGuid())
            .setCurrentView(currentViewBuilder)

        getClientEvents().dispatch(
            Dto.newBuilder()
                .setDirectedCommand(builder)
                .build()
        )
    }

    /**
     * Iterator of players
     *
     * @return
     */
    override fun iterator(): Iterator<Player> {
        return playerMap
            .values
            .iterator()
    }

    fun joined(event: PlayerJoinedEventDto) {
        val player = createOrGet(of(event.playerId))
        player.joined(event)
    }

    private fun left(playerId: EntityId) {
        val player = createOrGet(playerId)
        playerMap.remove(playerId)
        playerServerUpdateSet.remove(playerId)
        player.left()
    }

    fun consumeFromJoinTopic(domainEvent: DomainEvent) {
        if (domainEvent.event == null) {
            left(of(domainEvent.key))
        } else if (domainEvent.event.hasPlayerJoined()) {
            joined(domainEvent.event.getPlayerJoined())
        }
    }

    fun consumeFromPlayerActionTopic(domainEvent: DomainEvent) {
        if (domainEvent.event != null && domainEvent.event.hasPlayerMoved()) {
            get(domainEvent.key)?.moved(domainEvent.event.getPlayerMoved())
        }
        if (domainEvent.event != null && domainEvent.event.hasPlayerDestroyed()) {
            get(domainEvent.key)
                ?.let { p -> handleDestroyed(p, domainEvent.event.getPlayerDestroyed()) }
        }
        if (domainEvent.event != null && domainEvent.event.hasPlayerSpawned()) {
            get(domainEvent.key)?.spawned(domainEvent.event.getPlayerSpawned())
        }
    }

    fun handleDestroyed(player: Player, playerDestroyedEvent: PlayerDestroyedEventDto) {
        handleShockwave(player, playerDestroyedEvent.destroyedTimestamp)
        player.destroyed(playerDestroyedEvent)
    }

    /**
     * When a player is destroyed we generate a shockwave that pushes nearby players away
     * from the destroyed players last position.
     *
     * @param fromPlayer
     */
    fun handleShockwave(fromPlayer: Player, destroyedAt: Long) {
        this.physicsQueue.enqueue(shipExploded(fromPlayer, this, destroyedAt))
    }

    /**
     * This call marks the player is connected on this running instance.
     * This is used to update server calculated vector positions.
     *
     * @param playerId
     */
    fun connected(playerId: EntityId) {
        playerServerUpdateSet.add(playerId)
    }

    fun consumeFromPlayerFiredTopic(domainEvent: DomainEvent) {
        if (domainEvent.event != null && domainEvent.event.hasPlayerFired()) {
            val playerFiredEventDto = domainEvent.event.getPlayerFired()
            get(of(playerFiredEventDto.ownerPlayerId))?.fired(domainEvent.event.getPlayerFired())
        }
    }

    fun consumeFromBoltActionTopic(domainEvent: DomainEvent) {
        if (domainEvent.event != null && domainEvent.event.hasBoltExhausted()) {
            val event = domainEvent.event.getBoltExhausted()
            get(of(event.ownerPlayerId))?.boltExhausted()
        }
    }

    /**
     * Updates all connected player models' vector positions.
     *
     * @param timestamp the timestamp to update the players to.
     */
    fun fixedUpdate(timestamp: Long) {
        playerServerUpdateSet
            .asSequence()
            .map { key -> playerMap[key] }
            .forEach { p -> p?.fixedUpdate(timestamp) }
    }

    /**
     * Return the total number of players in the game.
     * Is weakly consistent and thread safe.
     *
     * @return
     */
    fun getTotalNumberOfPlayers(): Int {
        return this.playerMap.size
    }

    /**
     * Return the number of players connected this node.
     * Is weakly consistent and thread safe.
     *
     * @return
     */
    fun getNumberOfConnectedPlayers(): Int {
        return this.playerServerUpdateSet.size
    }

    fun hasConnectedPlayers(): Boolean {
        return this.playerServerUpdateSet.isNotEmpty()
    }

    fun getSpatialIndex(): PlayerSpatialIndex {
        return spatialIndexFactory.get()
    }

    fun isConnected(id: EntityId): Boolean {
        return playerServerUpdateSet.contains(id)
    }

    companion object {
        fun create(
            bolts: Bolts,
            clock: Clock,
            scoreBoard: ScoreBoard,
            barriers: Barriers,
            physicsQueue: PhysicsQueueEnqueue,
            spatialIndexFactory: PlayerSpatialIndexFactory
        ): Players {
            return Players(bolts, clock, scoreBoard, barriers, physicsQueue, spatialIndexFactory)
        }
    }
}
